src/bootstrap/builder.rs

```diff
@@ -1051,8 +1051,8 @@ impl<'a> Builder<'a> {
 
         cargo.env("RUSTC_VERBOSE", self.verbosity.to_string());
 
-        if self.config.deny_warnings {
-            cargo.env("RUSTC_DENY_WARNINGS", "1");
+        if self.config.deny_warnings && !mode.is_tool() {
```

Comment:
So this is a behavior change -- EXTERNAL_TOOL is not equivalent to !tool unfortunately. That'll mean that in-tree tools no longer get passed deny warnings and all.

Reply:
Agreed yeah, but we don't really have that many internal tools and putting `#![deny(warnings)]` on them as necessary didn't really seem like it was the end of the world.

Reply:
Hm, I think this PR doesn't add the deny(warnings) in to the internal tools though? I'd probably want to do that here since we don't want to regress on that.

Reply:
Sure, I've added a bunch of `#![deny(warnings)]` now

Reply:
Can't we track the is_external flag for tools in `builder.rs` to avoid regressing this?

Previously we had a single point (ok, two points) to which all the new lints could be added, now we have a bunch of `#![deny(warnings)]` annotations which are easy to miss and it's not immediately clear which crates should have them and which should not.

Reply:
It's possible, but it seems sort of overkill for this purpose. Most of the tools never change and don't really have any reason to follow the rest of the compiler religiously. They're built so rarely that having warnings isn't really a problem since they typically only show up in CI logs anyway.

Reply:
It's useful to have a single point to add flags affecting all Rust code in the codebase except for the code we don't control.
If you don't want to do it as a part of this PR, please make an issue to be fixed later.

Reply:
I personally think that for the most part just having deny(warnings) manually in the tools is fine but can definitely see the appeal of having this threaded through. Opened an issue: https://github.com/rust-lang/rust/issues/64523